src/ctc_core.c

```diff
@@ -0,0 +1,462 @@
+#include <stdlib.h>
+#include <regex.h>
+#include "ctc_core.h"
+
+pthread_once_t ctc_api_once_init = PTHREAD_ONCE_INIT;
+
+CTC_HANDLE ctc_pool[MAX_CTC_HANDLE_COUNT];
+
+void ctc_api_init (void)
+{
+    int i, j;
+
+    for (i = 0; i < MAX_CTC_HANDLE_COUNT; i ++)
+    {
+        ctc_pool[i].ID = i;
+        ctc_pool[i].control_session.session_gid = -1;
+
+        for (j = 0; j < MAX_JOB_HANDLE_COUNT; j ++)
+        {
+            ctc_pool[i].job_pool[j].ID = j;
+            ctc_pool[i].job_pool[j].job_session.job_desc = -1;
+        }
+    }
+}
+
+int alloc_ctc_handle (CTC_HANDLE **ctc_handle_p)
+{
+    int i;
+
+    *ctc_handle_p = NULL;
+
+    for (i = 0; i < MAX_CTC_HANDLE_COUNT; i ++)
+    {
+        if (ctc_pool[i].control_session.session_gid == -1)
+        {
+            *ctc_handle_p = &ctc_pool[i];
+            
+            break;
+        }
+    }
+
+    if (IS_NULL (*ctc_handle_p))
+    {
+        goto error;
+    }
+
+    return CTC_SUCCESS;
+
+error:
+
+    return CTC_FAILURE;
+}
+
+int free_ctc_handle (CTC_HANDLE *ctc_handle)
+{
+    // 소켓 정리, 할당된 job 정리
+
+    ctc_handle->control_session.session_gid = -1;
+
+    return CTC_SUCCESS;
+}
+
+int alloc_job_handle (CTC_HANDLE *ctc_handle, JOB_HANDLE **job_handle_p)
+{
+    int i;
+
+    *job_handle_p = NULL;
+
+    for (i = 0; i < MAX_JOB_HANDLE_COUNT; i ++)
+    {
+        if (ctc_handle->job_pool[i].job_session.job_desc == -1)
+        {
+            *job_handle_p = &ctc_handle->job_pool[i];
+            
+            break;
+        }
+    }
+
+    if (IS_NULL (*job_handle_p))
+    {
+        goto error;
+    }
+
+    return CTC_SUCCESS;
+
+error:
+
+    return CTC_FAILURE;
+}
+
+int free_job_handle (JOB_HANDLE *job_handle)
+{
+    job_handle->job_session.job_desc = -1;
+
+    return CTC_SUCCESS;
+}
+
+int find_ctc_handle (int ctc_handle_id, CTC_HANDLE **ctc_handle_p)
+{
+    CTC_HANDLE *ctc_handle;
+
+    ctc_handle = &ctc_pool[ctc_handle_id];
```

Comment:
One more example.
The value of MAX_CTC_HANDLE_COUNT is 100.
However, application can call ctc_api like this:
..
ctc_register_table(40000, 2, "dba", "t1");
  -> register_table (40000, 2, "dba", "t1");
     -> find_ctc_handle (40000, &ctc_hdndle);
        -> ctc_handle = &ctc_pool [40000];
You never checked integrity of 'ctc_handle_id' at any place.

Reply:
That is a bug. I will fix it.